common/test/TestLogging.cs

```diff
@@ -2,39 +2,114 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
 using System;
-using System.Diagnostics;
-using System.Globalization;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.Versioning;
+using Microsoft.ApplicationInsights;
+using Microsoft.ApplicationInsights.DataContracts;
+using Microsoft.ApplicationInsights.Extensibility;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
 
 namespace Microsoft.Azure.Devices.E2ETests
 {
+    /// <summary>
+    /// This class logs to multiple providers - Event source, Application insights
+    /// </summary>
     public class TestLogging
     {
-        private readonly static TestLogging s_instance = new TestLogging();
+        private readonly object initLock = new object();
+
+        // Client to log to application insights
+        private static TelemetryClient _telemetryClient;
+
+        // Unique ID for all tests of a run.
+        private static string _testRunId;
+
+        public TestContext Context { get; set; }
+        public string TargetFramework { get; set; }
 
         private TestLogging()
         {
+            if (_telemetryClient != null)
+            {
+                return;
+            }
+            else
+            {
+                // Thread-safe initialization of the telemetry client and run Id.
+                lock (initLock)
+                {
+                    string iKey = Environment.GetEnvironmentVariable("E2E_IKEY");
+                    if (!string.IsNullOrWhiteSpace(iKey))
```

Comment:
maybe log an exception if the key isn't set properly?

Reply:
We don't have the key in any prod key-vault yet so it will always log. After it starts working everywhere, I can just remove that check. I will add a task for it.

Reply:
no, I mean, in the scenario where something in our infra setup changed, and this key was not migrated over, there isn't any logging here to indicate what went wrong/ what is missing.

Reply:
There is no logging till a logger is setup anyway :) The only time I can throw an exception is if the value is null/empty (which I cannot do right now as it is not configured). What else are you suggesting to throw on?

Reply:
What my focus here is, there should be some way to inform us that since AI key isn't set up, nothing will get logged. It doesn't have to be an exception, a simple log statement would help too.

It is only to help us debug when we think that we have the infra set up as expected, but still don't see any records on application insights. Does that make sense?

Reply:
Ok then we are on the same page. Once I remove the if null check, AI will automatically throw the correct exception back if the key is null or empty, which will bubble up. We don't need to throw a different exception. 
It will throw this:
Initialization method Microsoft.Azure.Devices.E2ETests.BulkOperationsE2ETests.TestInitialize threw exception. System.ArgumentNullException: Value cannot be null.
Parameter name: InstrumentationKey.

However there is no way to know if you put in an incorrect iKey. This is better done by setting up an alert on AI itself. We can say send an alert if you see 0 logs from this service in a day.